pkg/fleet-manager/application/rollout_helper.go

```diff
@@ -0,0 +1,459 @@
+/*
+Copyright Kurator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package application
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	flaggerv1b1 "github.com/fluxcd/flagger/pkg/apis/flagger/v1beta1"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	applicationapi "kurator.dev/kurator/pkg/apis/apps/v1alpha1"
+	fleetapi "kurator.dev/kurator/pkg/apis/fleet/v1alpha1"
+	fleetmanager "kurator.dev/kurator/pkg/fleet-manager"
+	"kurator.dev/kurator/pkg/fleet-manager/plugin"
+)
+
+const (
+	RolloutLabel = "kurator.dev/rollout"
+)
+
+func (a *ApplicationManager) fetchRolloutClusters(ctx context.Context,
+	app *applicationapi.Application,
+	kubeClient client.Client,
+	originFleet *fleetapi.Fleet,
+	syncPolicy *applicationapi.ApplicationSyncPolicy) (map[fleetmanager.ClusterKey]*fleetmanager.FleetCluster, error) {
+
+	log := ctrl.LoggerFrom(ctx)
+	destination := getPolicyDestination(app, syncPolicy)
+	ClusterInterfaceList, result, err := a.fetchFleetClusterList(ctx, originFleet, destination.ClusterSelector)
```

Comment:
If you look closely, this is called in the caller, i would not calling it twice

Reply:
I will then refactor the fetch function of the destination cluster and create the client only once during the application's reconcile. Once the refactoring is complete, the cluster will not fetched twice.